L1Trigger/TrackerTFP/src/SeedFilter.cc

```diff
@@ -0,0 +1,242 @@
+    };
+    auto smallerChi2 = [this](const vector<StubSF*>& lhs, const vector<StubSF*>& rhs) {
+      auto chi2 = [this](const vector<StubSF*>& stubs) {
+        double d(0.);
+        for (StubSF* stub : stubs)
+          d += stub->chi2();
+        return d / ((int)stubs.size() - 2);
+      };
+      return chi2(lhs) < chi2(rhs);
+    };
+    auto moreLayers = [this](const vector<StubSF*>& lhs, const vector<StubSF*>& rhs) {
+      auto numLayers = [this](const vector<StubSF*>& stubs) {
+        TTBV hitPattern(0, setup_->numLayers());
+        for (StubSF* stub : stubs)
+          hitPattern.set(stub->layer());
+        return hitPattern.count();
+      };
+      return numLayers(lhs) > numLayers(rhs);
+    };
+    const int numLayers = 16;
+    static const vector<pair<int, int>> seedingLayers = {{1, 2}, {1, 3}, {2, 3}, {1, 11}, {2, 11}, {1, 12}, {2, 12}, {11, 12}};
+    const double dT = setup_->chosenRofPhi() - setup_->chosenRofZ();
+    for (int channel = 0; channel < dataFormats_->numChannel(Process::mht); channel++) {
+      TTDTC::Stream& acceptedStream = accepted[region_ * dataFormats_->numChannel(Process::mht) + channel];
+      TTDTC::Stream& lostStream = lost[region_ * dataFormats_->numChannel(Process::mht) + channel];
+      deque<StubSF*> output;
+      vector<StubMHT*>& stubs = input_[channel];
+      stubs.erase(remove(stubs.begin(), stubs.end(), nullptr), stubs.end());
+      set<int> ids;
+      for (StubMHT* stub : stubs)
+        if (stub)
+          ids.insert(stub->trackId());
+      vector<vector<StubMHT*>> tracks(ids.size());
+      int i(0);
+      for (auto it = stubs.begin(); it != stubs.end();) {
+        const auto start = it;
+        const int id = (*it)->trackId();
+        auto different = [id](StubMHT* stub){ return id != stub->trackId(); };
+        it = find_if(it, stubs.end(), different);
+        vector<StubMHT*>& track = tracks[i++];
+        track.reserve(distance(start, it));
+        copy(start, it, back_inserter(track));
+      }
+      for (const vector<StubMHT*>& track : tracks) {
+        const int eta = track.front()->sectorEta();
+        const double dZT = (sinh(setup_->boundarieEta(eta + 1)) - sinh(setup_->boundarieEta(eta))) * setup_->chosenRofZ();
+        vector<vector<StubMHT*>> layerStubs(numLayers);
+        for (vector<StubMHT*>& layer : layerStubs)
+          layer.reserve(track.size());
+        for (StubMHT* stub : track)
+          layerStubs[layerId(stub)].push_back(stub);
+        deque<pair<StubMHT*, StubMHT*>> seeds;
+        for (const pair<int, int>& seedingLayer : seedingLayers)
+          for (StubMHT* inner : layerStubs[seedingLayer.first])
+            if (setup_->psModule(inner->ttStubRef()))
+              for (StubMHT* outer : layerStubs[seedingLayer.second])
+                if (setup_->psModule(outer->ttStubRef()))
+                  if (outer->r() > inner->r())
+                    seeds.emplace_back(inner, outer);
+        vector<vector<StubSF*>> seedTracks;
```

Comment:
If you don't want a space (empty line) to separate significant elements of the logic, add a comment. e.g. "Verify which seeds form good track candidates"